neqo-transport/src/connection.rs

```diff
@@ -1028,41 +1035,89 @@ impl Connection {
             let mut tokens = Vec::new();
 
             // Ensure we have tx crypto state for this epoch, or skip it.
-            if !matches!(
-                self.crypto.obtain_crypto_state(self.role, epoch),
-                Ok(CryptoState { tx: Some(_), .. })
-            ) {
-                continue;
-            }
+            let tx = match self
+                .crypto
+                .states
+                .obtain(self.role, epoch, &self.crypto.tls)
+            {
+                Ok(CryptoState { tx: Some(tx), .. }) => tx,
+                _ => continue,
+            };
+
+            let hdr = PacketHdr::new(
+                0,
+                match epoch {
+                    0 => {
+                        let token = match &self.retry_info {
+                            Some(v) => v.token.clone(),
+                            _ => Vec::new(),
+                        };
+                        PacketType::Initial(token)
+                    }
+                    1 => {
+                        if self.zero_rtt_state == ZeroRttState::Rejected {
+                            continue;
+                        }
+                        self.zero_rtt_state = ZeroRttState::Sending;
+                        PacketType::ZeroRTT
+                    }
+                    2 => PacketType::Handshake,
+                    3 => PacketType::Short,
+                    _ => unimplemented!(), // TODO(ekr@rtfm.com): Key Update.
+                },
+                Some(self.version),
+                path.remote_cid.clone(),
+                path.local_cids.first().cloned(),
+                self.loss_recovery.next_pn(space),
+                epoch,
+            );
 
             let mut ack_eliciting = false;
+            let tx_mode = self.tx_mode;
+
             match &self.state {
                 State::Init | State::WaitInitial | State::Handshaking | State::Connected => {
                     loop {
-                        let remaining = path.mtu() - out_bytes.len() - encoder.len();
-
-                        // Check sources in turn for available frames
-                        if let Some((frame, token)) = self
-                            .acks
-                            .get_frame(now, epoch)
-                            .or_else(|| self.crypto.get_frame(epoch, TxMode::Normal, remaining))
-                            .or_else(|| self.flow_mgr.borrow_mut().get_frame(epoch, remaining))
-                            .or_else(|| {
-                                self.send_streams
-                                    .get_frame(epoch, TxMode::Normal, remaining)
-                            })
-                        {
+                        let used =
+                            out_bytes.len() + encoder.len() + hdr.overhead(&tx.aead, path.mtu());
+                        let remaining = path.mtu() - used;
+                        if remaining == 0 {
```

Comment:
Maybe we should acknowledge the fact that all useful frames are at least 2 bytes:

```suggestion
                        if remaining <= 1 {
```